scipy/spatial/_hausdorff.pyx

```diff
@@ -22,26 +20,22 @@ __all__ = ['directed_hausdorff']
 @cython.boundscheck(False)
 def directed_hausdorff(double[:,::1] ar1, double[:,::1] ar2, seed=0):
 
-    cdef double cmax, cmin, d
+    cdef double cmax, cmin, d = 0
     cdef int N1 = ar1.shape[0]
     cdef int N2 = ar2.shape[0]
     cdef int data_dims = ar1.shape[1]
-    cdef unsigned int i, j, k
+    cdef int i, j, k
     cdef unsigned int i_store = 0, j_store = 0, i_ret = 0, j_ret = 0
-    cdef long[:] resort1, resort2
+    cdef np.ndarray[np.int64_t, ndim=1, mode='c'] resort1, resort2
```

Comment:
Just for me to understand: this was supposed to be an 'old syntax', to be replaced by typed memoryviews where possible/reasonable? If numpy 1.20 changes this, can we spell a recommendation somewhere, as to which syntax to prefer.

The original is a wee bit strange, too: why it's `long[:]` not `long[::1]`, if it's basically an arange, so it's contiguous.

Reply:
It's the declaration for the variable holding the result of `np.arange`, so yes this is supposed to be old syntax (see gh-13470).  And we need an `ndarray` here, because `shuffle` throws a warning for memoryviews in numpy 1.20

Reply:
Can't NumPy just not issue a warning for cases where there's no actual risk? I'm still struggling to understand why the burden is on us to make a change here.

It may be that it would be much harder on the NumPy side--you'd have to inspect to see that the memoryview is 1-dimensional C-contiguous and then make the decision, but still, there's no bug and it seems like we're kind of getting pressured to backport a fix for a non-bug.

Reply:
I've proposed it upstream at least: https://github.com/numpy/numpy/issues/18273 

Certainly if the NumPy team were open to it I'd like to see the scientific code changes kept to a minimum here, perhaps even just leaving the function completely alone (reverting the warning shim) after NumPy `1.20.1`.

If there are genuinely strong reasons for the NumPy team to consider the 1-D C-contiguous case something other than a bystander, then ok yeah probably the warning shim is not ideal long term.

Reply:
The changes here are straightforward, so I'm not sure why we wouldn't backport them. There were (are) silent correctness issues for `shuffle`, where it would make assumption about `object` inputs that could be anything, and may be wrong. The warning should not be reverted.

Reply:
> The changes here are straightforward, so I'm not sure why we wouldn't backport them. 

Because they're not needed if NumPy special cases the 1-D C-contiguous memoryviews. I mean that we revert the warning shim in our code once NumPy is more targeted in raising their warning, so that we ultimately don't change anything on our end.

Reply:
In this case, we can set aside the `shuffle()` warning from consideration. `resort1` and `resort2` simply shouldn't be memoryviews. They start life as `ndarray`s, are passed to a function that is primarily meant for and has a fast path for `ndarray`s, then are used as fancy indices which also are expected to be `ndarray`s. Using `cdef long[:]" here was always suboptimal (not wrong, just suboptimal). This is a positive change regardless of what `shuffle()` warns on.

Reply:
Ok, we could reframe it as an enhancement if you think this will genuinely help performance (I can try to check at some point..) or overhead or whatever, which wouldn't normally be backported either.

Reply:
numpy 1.20.0 is released and will be in the valid range for the scipy dependency on it for a good long time, so avoiding the spurious warning is a reasonable thing to backport to a bugfix release. There's no need to undo it ever, even if the original motivation goes away, because it's a positive change (if just by a hair), not a cumbersome workaround.

Reply:
Ok fair enough, at least we agree the warning is spurious--it would probably make a fair bit more work for me to try to selectively set`!=1.20.0` etc. in a release process anyway.

If the old buffer syntax gets deprecated maybe it could change a bit.